### DISCUSSION TITLE
Compare riscv-unified-db schema for CSRs to riscv-config

### DISCUSSION THREAD

**Discussion by james-ball-qualcomm** [riscv-software-src/riscv-unified-db#197] [General]
See https://riscv-config.readthedocs.io/en/latest/yaml-specs.html#csr-template

**Comment by allenjbaum** [comment]
I don't think I've ever seen the schema for CSRs - where is it?

**Comment by dhower-qc** [comment]
The enforced schema is here:

https://github.com/riscv-software-src/riscv-unified-db/blob/main/schemas/csr_schema.json

Examples are under arch/csr:

https://github.com/riscv-software-src/riscv-unified-db/tree/main/arch/csr

**Comment by Ganesh-403** [comment]
If you're looking to compare the two formats, the key references are:

- **RISCV-Config CSR template** (linked in the discussion), which defines a configuration-oriented schema for describing CSRs, including RV32/RV64 variants, field accessibility, reset values, and WARL/WLRL semantics. :contentReference[oaicite:0]{index=0}
- **riscv-unified-db's CSR schema**, implemented as `schemas/csr_schema.json`, with example CSR definitions under `arch/csr/`. As Derek pointed out, this is the authoritative schema enforced by UDB.

The two projects have different goals, which is why their schemas are not identical:

- **RISCV-Config** is primarily a **configuration validator**. Its CSR YAML is designed to describe a *specific implementation*, including whether fields are implemented, accessibility in different XLENs, reset values, and legal WARL behavior. :contentReference[oaicite:1]{index=1}
- **riscv-unified-db** aims to be a **machine-readable representation of the RISC-V specifications**. Beyond structural information about CSRs, it also captures specification metadata, normative prose, semantic information, versioning, and relationships between ISA components so that documentation and other artifacts can be generated from a single source. :contentReference[oaicite:2]{index=2}

Because of that, there isn't necessarily a one-to-one mapping between the two schemas. UDB is modeling the specification itself, whereas RISCV-Config is modeling a concrete processor configuration against that specification.

If the goal of this discussion is interoperability, it would be useful to identify:

- fields that map directly between the two formats,
- information present in RISCV-Config but missing in UDB,
- information present in UDB but intentionally absent from RISCV-Config due to their different scopes.

That kind of comparison could help determine whether compatibility or conversion tooling would be valuable without forcing either project to adopt the other's schema wholesale.

If this solves your problem, feel free to mark it as the accepted answer so others can find it easily.

**Reply by allenjbaum** [reply]
Both RISCV-config and UDB are machine readable descriptions. There is also a SAIL schema to consider.  
But: the schema is the outline. What we actually use for certification is the implementation specif configuration, which must be a proper subset of any schema. I'm not sure there is a lot in RISCV-config that isn't in UDB, although there is a bunch of WARL behavior that can probably be deprecated or changed (specifically, the illegal->legal mappings which are incomplete (completeness is literally not possbile, and we aren't testing that at all). I think the biggest difference is the idea of conditional behaviors (so, different WARL behaviors depending on the values of either state bits (in CSRs) or parameters (asserted in the implementation defined YAML or JSON)
